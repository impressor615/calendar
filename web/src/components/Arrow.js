import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';


const Arrow = ({ direction, onClick, ...rest }) => (
  <button
    {...rest}
    onClick={onClick}
    className={classnames(
      'arrow',
      rest.className,
      { [direction]: direction },
    )}
  />
);

Arrow.defaultProps = {
  direction: 'previous',
};

Arrow.propTypes = {
  direction: PropTypes.oneOf(['next', 'previous']),
  onClick: PropTypes.func.isRequired,
};

export default Arrow;