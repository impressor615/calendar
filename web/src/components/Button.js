import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';


const Button = ({ children, color, onClick, active, ...rest }) => (
  <button
    type="button"
    onClick={onClick}
    {...rest}
    className={classnames(
      'btn',
      rest.className,
      { [color]: color },
      { active },
    )}>
    { children }
  </button>
);

Button.defaultProps = {
  color: '',
  active: false,
  onClick: () => {},
};

Button.propTypes = {
  color: PropTypes.oneOf(['', 'primary', 'secondary']),
  onClick: PropTypes.func,
  children: PropTypes.node.isRequired,
  active: PropTypes.bool,
}

export default Button;
