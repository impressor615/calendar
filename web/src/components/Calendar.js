import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import CONSTANTS from '../constants';


const { DAY, HOURS } = CONSTANTS;
export const CalendarDay = ({
  day,
  date,
  children,
  onToggle,
}) => (
  <button className="cal-block" onClick={onToggle}>
    { day ? <div className="day">{day}</div> : null}
    <div>{date}</div>
    {children}
  </button>
);

export const MonthCalendar = ({ range, onToggle }) => (
  <section className="month-calendar">
    {
      range.map((date, index) => (
        <Fragment key={index.toString()}>
          {
            index <= 6
              ? <CalendarDay date={date} day={DAY[index]} onToggle={onToggle} />
              : <CalendarDay date={date} onToggle={onToggle} />
          }
        </Fragment>
      ))
    }
  </section>
);

export const CalendarTime = ({ children, onToggle }) => (
  <button className="cal-block sm" onClick={onToggle}>
    {children}
  </button>
);

export const WeekCalendar = ({ range, onToggle }) => (
  <section className="week-calendar">
    <div className="cal-block sm" />
    {
      range.map((date, index) => (
        <div className="cal-block sm" key={index.toString()}>
          <div className="day">{DAY[index]}</div>
          <div>{date}</div>
        </div>
      ))
    }
    {
      HOURS.map((hour, index) => (
        <Fragment key={index.toString()}>
          <div className="cal-block sm">
            <span className="hour">{hour}</span>
          </div>
          {
            range.map((date, index) => (
              <CalendarTime key={index.toString()} onToggle={onToggle} />
            ))
          }
        </Fragment>
      ))
    }
  </section>
);

const Calendar = ({ type, range, onToggle }) => (
  <article className="calendar">
    {
      type === 'month'
        ? <MonthCalendar range={range} onToggle={onToggle} />
        : <WeekCalendar range={range} onToggle={onToggle} />
    }
  </article>
);

Calendar.propTypes = {
  type: PropTypes.string.isRequired,
  range: PropTypes.arrayOf(PropTypes.number).isRequired,
  onToggle: PropTypes.func.isRequired,
};

export default Calendar;
