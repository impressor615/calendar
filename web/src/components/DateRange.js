import React from 'react';
import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';

const DateRange = ({ startDate, endDate, onChange }) => (
  <div className="date-range">
    <DatePicker
      showTimeSelect
      selected={startDate}
      onChange={onChange}
      timeIntervals={60}
      timeCaption="time"
      dateFormat="YYYY-M-D h:mm A"
    />
    <div>~</div>
    <div>{ endDate.format('YYYY-M-D h:mm A') }</div>
  </div>
);

DateRange.propTypes = {
  startDate: PropTypes.object.isRequired,
  endDate: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
}

export default DateRange;
