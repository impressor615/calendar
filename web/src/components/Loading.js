import React from 'react';
import PropTypes from 'prop-types';

import Modal, { ModalDialog } from 'components/Modal';



const Loading = ({ isOpen, children }) => (
  <Modal isOpen={isOpen}>
    <ModalDialog className="center">
      {children}
    </ModalDialog>
  </Modal>
);

Loading.defaultProps = {
  children: 'Loading...',
};

Loading.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  children: PropTypes.node,
};

export default Loading;
