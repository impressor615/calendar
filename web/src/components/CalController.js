import React from 'react';
import PropTypes from 'prop-types';

import { withContext } from 'AppContext';
import Arrow from 'components/Arrow';
import Button from 'components/Button';
import { getCalendarTitle } from 'utils/dateUtils';


const CalController = ({
  type,
  currentDate,
  onArrowClick,
  onTypeChange,
}) => (
  <article className="controller">
    <div className="controller-wrapper">
      <section className="date-control">
        <Arrow name="previous" onClick={onArrowClick} />
        <div className="title">{getCalendarTitle(currentDate, type)}</div>
        <Arrow name="next" direction="next" onClick={onArrowClick} />
      </section>
      <section className="type-control">
        <Button name="month" onClick={onTypeChange} color="primary" active={type === 'month'}>월</Button>
        <Button name="week" onClick={onTypeChange} color="primary" active={type === 'week'}>주</Button>
      </section>
    </div>
  </article>
);

CalController.propTypes = {
  type: PropTypes.string.isRequired,
  currentDate: PropTypes.object.isRequired,
  onArrowClick: PropTypes.func.isRequired,
  onTypeChange: PropTypes.func.isRequired,
}

export default withContext(CalController);
export const PureCalController = CalController;
