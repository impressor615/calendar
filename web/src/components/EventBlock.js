import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Button from 'components/Button';


const EventBlock = ({
  children,
  onToggle,
  onDragStart,
  fluid,
  ...rest
}) => (
  <Button
    {...rest}
    draggable
    onClick={onToggle}
    onDragStart={onDragStart}
    className={classnames(
      'event-block',
      rest.className,
      { fluid },
    )}
  >
    { children }
  </Button>
);

EventBlock.defaultProps = {
  fluid: false,
};

EventBlock.propTypes = {
  children: PropTypes.node.isRequired,
  onToggle: PropTypes.func.isRequired,
  onDragStart: PropTypes.func.isRequired,
  fluid: PropTypes.bool,
}

export default EventBlock;