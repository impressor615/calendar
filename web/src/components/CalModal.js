import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { withContext } from 'AppContext';
import Modal, { ModalDialog, ModalHeader, ModalBody, ModalFooter } from 'components/Modal';
import DateRange from 'components/DateRange';
import Button from 'components/Button';
import { postEvent, deleteEvent, updateServerEvent } from 'utils/fetchUtils';
import { getErrorMsg } from 'utils/errorUtils';
import CONSTANTS from '../constants';


const { NOTIFICATION } = CONSTANTS;
class CalModal extends Component {
  constructor(props) {
    super(props);
    this.onUpdate = this.onUpdate.bind(this);
    this.onDelete = this.onDelete.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onUpdate(e) {
    e.preventDefault();
    const {
      event,
      events,
      notify,
      setLoading,
      onToggle,
      updateEvents,
    } = this.props;
    const { _id, title, start_date, end_date } = event;
    if(!_id || !title || !start_date || !end_date) {
      return;
    }

    const putData = {
      title,
      start_date: start_date.toISOString(),
      end_date: end_date.toISOString(),
    };
    setLoading();
    updateServerEvent(_id, putData, events)
      .then((newEvents) => {
        setLoading();
        updateEvents(newEvents)
        onToggle();
        notify(({
          type: 'success',
          message: NOTIFICATION.update,
        }));
      }, (error) => {
        setLoading();
        const message = getErrorMsg(error)
        this.notify({ message });
      });
  }


  onDelete(e) {
    e.preventDefault();
    if (!window.confirm('정말 삭제하시겠습니까?')) {
      return;
    }

    const {
      event,
      events,
      notify,
      setLoading,
      updateEvents,
      onToggle,
    } = this.props;
    const { _id } = event;
    if (!_id) {
      return;
    }

    setLoading();
    deleteEvent(_id, events)
      .then((newEvents) => {
        setLoading();
        updateEvents(newEvents);
        onToggle();
        notify({
          type: 'success',
          message: NOTIFICATION.delete,
        });
      }, (error) => {
        setLoading();
        const message = getErrorMsg(error);
        notify({ message });
      });
  }

  onSubmit(e) {
    e.preventDefault();
    const {
      event,
      events,
      notify,
      onToggle,
      setLoading,
      updateEvents,
    } = this.props;
    const { title, start_date, end_date } = event;
    const postData = {
      title,
      start_date: start_date.toISOString(),
      end_date: end_date.toISOString(),
    };
    setLoading();
    postEvent(postData, events)
      .then((newEvents) => {
        setLoading();
        updateEvents(newEvents);
        onToggle();
        notify({
          type: 'success',
          message: NOTIFICATION.create,
        });
      }, (error) => {
        setLoading();
        const message = getErrorMsg(error);
        notify({ message });
      });
  }

  render() {
    const {
      isOpen,
      isLoading,
      onToggle,
      onChange,
      onDateChange,
      event,
    } = this.props;
    const { _id } = event;
    const onFormSubmit = _id ? this.onUpdate : this.onSubmit;
    return (
      <Modal
        isOpen={isOpen}
        onToggle={onToggle}
        className="cal-modal"
      >
        <ModalDialog>
          <ModalHeader onToggle={onToggle}>
            일정 추가 및 변경
          </ModalHeader>
          <form onSubmit={onFormSubmit}>
            <ModalBody>
              <input
                id="title"
                type="text"
                placeholder="일정 제목을 입력해주세요!"
                onChange={onChange}
                value={event.title}
                required
              />
              <DateRange
                startDate={event.start_date}
                endDate={event.end_date}
                onChange={onDateChange}
              />
            </ModalBody>
            <ModalFooter>
              <Button onClick={onToggle} className="close-btn" color="primary">취소</Button>
              { event._id ? <Button onClick={this.onDelete} className="delete-btn" color="primary">삭제</Button> : null }
              <Button type="submit" disabled={isLoading} className="submit-btn" color="secondary">저장</Button>
            </ModalFooter>
          </form>
        </ModalDialog>
      </Modal>
    );
  }
}

CalModal.propTypes = {
  onChange: PropTypes.func.isRequired,
  onDateChange: PropTypes.func.isRequired,
  event: PropTypes.object.isRequired,
  isOpen: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  setLoading: PropTypes.func.isRequired,
  onToggle: PropTypes.func.isRequired,
  notify: PropTypes.func.isRequired,
  updateEvents: PropTypes.func.isRequired,
  events: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default withContext(CalModal);
export const PureCalModal = CalModal;
