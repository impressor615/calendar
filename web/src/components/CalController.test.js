import React from 'react';
import moment from 'moment';

import Arrow from 'components/Arrow';
import Button from 'components/Button';
import { PureCalController as CalController } from 'components/CalController';


describe('<CalController />', () => {
  const onArrowClick = jest.fn();
  const onTypeChange = jest.fn();
  const defaultDate = moment('2018-10-13');
  const props = {
    currentDate: defaultDate,
    type: 'months',
    onArrowClick,
    onTypeChange,
  };

  it('should be rendered properly', () => {
    const wrapper = shallow(<CalController {...props} />);
    expect(wrapper.find('.date-control')).toHaveLength(1);
    expect(wrapper.find(Arrow)).toHaveLength(2);
    expect(wrapper.find('.date-control .title')).toHaveLength(1);
    expect(wrapper.find('.type-control')).toHaveLength(1);
    expect(wrapper.find(Button)).toHaveLength(2);
  });

  it('methods should be fired properly', () => {
    const wrapper = shallow(<CalController {...props} />);
    wrapper.find(Arrow).first().simulate('click');
    wrapper.find(Button).first().simulate('click');

    expect(onArrowClick).toHaveBeenCalledTimes(1);
    expect(onTypeChange).toHaveBeenCalledTimes(1);
  });

  it('should be matched with the snapshot', () => {
    const wrapper = mount(<CalController {...props} />);
    expect(wrapper).toMatchSnapshot();
  })
})