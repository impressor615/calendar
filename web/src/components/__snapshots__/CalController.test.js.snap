// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalController /> should be matched with the snapshot 1`] = `
<CalController
  onArrowClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  onTypeClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  title="title"
  type="months"
>
  <article
    className="controller"
  >
    <section
      className="date-control"
    >
      <button
        className="btn arrow left"
        name="previous"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        type="button"
      />
      <div
        className="title"
      >
        title
      </div>
      <button
        className="btn arrow right"
        name="next"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        type="button"
      />
    </section>
    <section
      className="type-control"
    >
      <button
        className="btn secondary"
        name="month"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        type="button"
      >
        월
      </button>
      <button
        className="btn secondary"
        name="week"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
        type="button"
      >
        주
      </button>
    </section>
  </article>
</CalController>
`;
